python/cudf/cudf/core/scalar.py

```diff
@@ -1,26 +1,208 @@
 # Copyright (c) 2020-2025, NVIDIA CORPORATION.
 from __future__ import annotations
 
+import copy
 import decimal
 import functools
 import operator
 from collections import OrderedDict
+from typing import TYPE_CHECKING, Any
 
 import numpy as np
+import pandas as pd
 import pyarrow as pa
 
 import pylibcudf as plc
 
 import cudf
 from cudf.api.types import is_scalar
-from cudf.core.dtypes import ListDtype, StructDtype
+from cudf.core.dtypes import (
+    Decimal32Dtype,
+    Decimal64Dtype,
+    ListDtype,
+    StructDtype,
+)
 from cudf.core.missing import NA, NaT
 from cudf.core.mixins import BinaryOperand
 from cudf.utils.dtypes import (
     get_allowed_combinations_for_operator,
     to_cudf_compatible_scalar,
 )
 
+if TYPE_CHECKING:
+    from typing_extensions import Self
+
+    from cudf._typing import Dtype, ScalarLike
+
+
+def _preprocess_host_value(value, dtype) -> tuple[ScalarLike, Dtype]:
```

Comment:
I feel like this function can definitely be simplified, or at least optimized to reduce the complexity of the conditional cascades, but given that this is not a new function we can do that later and keep this PR focused on the DeviceScalar removal.